content/blog/2019-09-25-floats-static-arrays.md

```diff
@@ -0,0 +1,67 @@
+
+In Bril, floating-point precisions should not be mixed.  It is expected, for instance, that the following code will produce a static exception:
+```
+v0: double = const 5;
+v1: float = const 5;
+v2: float = fadd v0 v1
+```
+Similarly, mixing floating-points and integers in Bril is undefined, as well as using floating-point arithmetic operations on integers or vice-versa.  In a future project, I intend to explore the consequences of relaxing these semantic requirements, particularly when mixing floating-point precision.
+
+### Bril Fixed-Length Arrays
+
+Bril now supports types of the form `array=type[size]`, where `type = | int | bool | float | double | array`.  Note the inclusion of type recursion in this definition, which permits multi-dimensional arrays.  Arrays in Bril must contain elements all of the same type, and the size of arrays cannot be changed after initialization.
+
+There are three operations related to manipulating arrays in Bril: initialization, assignment, and indexing.  Arrays are initialized with the `new` instruction, which takes a type and stores a default initialization of that type to the given destination.  Array elements can be set with the `set` instruction, an effectful instruction which takes the array, an index, and a value in that order.  Finally, array elements can be read using the `index` instruction, which writes a value of the given array element.
+
+Array operations are concisely summarized in the following code snippet:
+```
+val: int = const 5;
+ind: int = const 1;
+
+arr: int[3] = new int[3];   // arr = [0, 0, 0]
+set arr ind val;            // arr = [0, 5, 0]
+res: int = index arr ind;   // res = 5
+```
+Arrays can be written to freely in any Bril block.  As a result, array values are inherently stateful and should be treated by users carefully.  Writing values of the wrong type, however, is expected to not be permitted.
+
+### Implementation
+
+Floating points and fixed-length arrays have been implemented as described, and can be reasoned about by the `bril2json` parser, the `Brili` interpreter, and the `bril2txt` translator.  A subset of TypeScript can be written to Bril-recognized JSON by the `ts2bril` compiler.  While these conversions work generally as described above, there are some features which bear special notice.
+
+The `Brili` interpreter supports integers, doubles, and floating-points all as different objects.  Integers are represented as Javascript [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) objects.  This choice of representation means that `Brili` does not correctly reason about integer overflow -- while this should be implemented in the future, I have not observed a need for it in the current Bril projects.
```

Comment:
```suggestion
The `Brili` interpreter supports integers, doubles, and floating-points all as different objects.  Integers are represented as JavaScript [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) objects.  This choice of representation means that `Brili` does not correctly reason about integer overflow -- while this should be implemented in the future, I have not observed a need for it in the current Bril projects.
```

Reply:
Indeed; I haven't even specified yet what's supposed to happen on integer overflow!